14110 - Expose height and width attributes to 'wp_get_attachment_image_attributes' filter

spacedmonkey opened the pull request:
<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

<!-- Insert a description of your changes here -->

Trac ticket: https://core.trac.wordpress.org/ticket/14110

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

spacedmonkey committed:
Ensure width and height attributes are sanitized for images

Added validation to sanitize 'width' and 'height' attributes before use and updated the default attributes accordingly. This improves security and prevents potential issues with invalid values in generated image markup.

spacedmonkey committed:
Optimize image loading attribute handling in `wp_get_attachment_image`.

Simplified the code by directly passing `$attr` instead of creating and assigning a redundant `$loading_attr` array. This improves readability and eliminates unnecessary variable usage while maintaining functionality.

spacedmonkey committed:
Normalize image attributes before unset

Refactored the handling of `height` and `width` attributes by storing their sanitized values in separate variables before unsetting them. This aims to improve code readability and maintainability.

spacedmonkey committed:
Ensure height and width attributes are numeric before sanitizing.

Previously, the height and width attributes were directly sanitized using `absint`, which could lead to unexpected behavior if non-numeric values were passed. This update adds a check to confirm the values are numeric before sanitizing, improving data validation and preventing potential errors.

spacedmonkey committed:
Simplify height and width attribute handling in media.php.

Consolidated logic for processing height and width attributes by combining conditions and reducing redundant checks. Unset operations for these attributes are now grouped for clarity. This improves code readability and maintainability.

adamsilverstein commented on src/wp-includes/media.php:
```diff
@@ -1169,8 +1167,16 @@ function wp_get_attachment_image( $attachment_id, $size = 'thumbnail', $icon = f
+		if ( isset( $attr['width'] ) && is_numeric( $attr['width'] ) ) {
+			$width = absint( $attr['width'] );
+		}
+		unset( $attr['height'], $attr['width'] );
```
Curious why we need to unset these here?

felixarntz replied:
I think otherwise `width` and `height` would be included in the attributes twice, per the `$attr` loop below.

spacedmonkey replied:
That is correct, we don't want to the height and width twice.

adamsilverstein reviewed (approved):
Looks great! One small question.

felixarntz commented on src/wp-includes/media.php:
```diff
@@ -1070,17 +1070,18 @@ function wp_get_attachment_image( $attachment_id, $size = 'thumbnail', $icon = f
+			'src'    => $src,
+			'class'  => "attachment-$size_class size-$size_class",
+			'alt'    => trim( strip_tags( get_post_meta( $attachment_id, '_wp_attachment_image_alt', true ) ) ),
+			'width'  => $width,
+			'height' => $height,
```
I don't think we should merge `width` and `height` into `$default_attr`. Doing that would allow overriding their values via `$attr`, which I don't think we should allow without further consideration.

I think the better solution would be to leave `$default_attr` untouched, and instead only add the two values to `$attr` _after_ the `wp_parse_args( $attr, $default_attr )` call.

spacedmonkey replied:
Fixed

felixarntz commented on src/wp-includes/media.php:
```diff
@@ -1093,12 +1094,9 @@ function wp_get_attachment_image( $attachment_id, $size = 'thumbnail', $icon = f
 		$context = apply_filters( 'wp_get_attachment_image_context', 'wp_get_attachment_image' );
 		$attr    = wp_parse_args( $attr, $default_attr );
```
See my comment above: Better set the two values in `$attr` here instead of modifying `$default_attr`.
```suggestion
		$context        = apply_filters( 'wp_get_attachment_image_context', 'wp_get_attachment_image' );
		$attr           = wp_parse_args( $attr, $default_attr );
		$attr['width']  = $width;
		$attr['height'] = $height;
```

spacedmonkey replied:
Good catch @felixarntz . Fixed in https://github.com/WordPress/wordpress-develop/pull/8758/commits/53776df91dcb72a8912a4a2daa0b5d34d73d8043

felixarntz reviewed (dismissed):
@spacedmonkey Mostly looks good to me, except it would be safer to modify the `$attr` array directly.

spacedmonkey committed:
Merge branch 'trunk' into fix/height-width

spacedmonkey committed:
Merge branch 'trunk' into fix/height-width

spacedmonkey committed:
Include `width` and `height` in `$attr` after merging with defaults.

Ensured `width` and `height` attributes are always set in the final `$attr` array for consistent handling of image dimensions. Improved code readability by aligning array keys.

felixarntz reviewed (commented):
@spacedmonkey Production code looks good.

Would be great to add a test case where the `width` and/or `height` are modified via filter, to ensure this actually works.

spacedmonkey committed:
Merge branch 'trunk' into fix/height-width

spacedmonkey committed:
Add PHPUnit tests for `wp_get_attachment_image` attribute filters

Introduced tests to validate the behavior of `wp_get_attachment_image_attributes` filter, ensuring proper handling of `width` and `height` attributes. Includes tests for default values, modified values, and unset scenarios to ensure consistent attribute handling.
